### PR TITLE
Add devise

### PR DESCRIPTION
**Pull request by MayPyone**
- Remove current_user method in ApplicationController, devise will provide us with one.
- [Install and setup devise](https://github.com/heartcombo/devise#getting-started).
- Can register a new user.
- User logs in with a combination of email and password.
- Hashed password should be stored in the database.
- Ask for confirmation of email.
- Can reset password.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+  before_action :update_allowed_parameters, if: :devise_controller?
+
+  protected
+
+  def update_allowed_parameters
+    devise_parameter_sanitizer.permit(:sign_up) do |user_params|
+      user_params.permit({ roles: [] }, :email, :password, :name, :posts_counter)
+    end
+    devise_parameter_sanitizer.permit(:account_update) do |user_params|
+      user_params.permit({ roles: [] }, :email, :name, :password, :current_password)
```
Comment (topeogunleye) [inline app/controllers/application_controller.rb:13]: Hi, thank you for submitting your code. I have reviewed your application_controller.rb file and here are some feedbacks:

- Your code is well-organized and easy to read. You have used before_action and devise_parameter_sanitizer correctly to handle authentication and authorization. 👍

### REQUIRED CHANGES ♻️ 
- You have not permitted the password_confirmation parameter for the sign_up and account_update actions. This is causing the error message of Unpermitted parameter. :password_confirmation. This parameter is required by Devise to validate the password and confirm that the user has entered the same password twice. You might want to add this parameter to the list of permitted parameters for both actions. 😕

image

Overall, you have done a good job with your code. I hope you find these feedbacks helpful and improve your code accordingly. If you have any other questions, please feel free to ask me. 😊

File: app/models/user.rb [hunk 2]
```diff
@@ -1,4 +1,8 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable, :confirmable
```
Comment (TedLivist) [inline app/models/user.rb:5]: - [ ] When I sign up the app raises an error. Kindly fix this. I also observed your schema and can see there is no `confirmation_token` on the user table, so users cannot confirm their emails after signing up.

  image

  image

- [ ] Kindly go through your app look closely at the requirements and ensure everything works.

File: app/views/devise/registrations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Sign up" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (TedLivist) [inline app/views/devise/registrations/new.html.erb:30]: - [ ] **[OPTIONAL]** I would recommend you center your sign up form like we have in the wireframe.

  image

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @MayPyone and @paulshaun-kdtk,

✔️ Installation of Devise

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/registrations/new.html.erb (hunk 3 above)
Comment (TedLivist) [reply app/views/devise/registrations/new.html.erb:30]: That's interesting. I set up your project again, same issue.

Could you clone the github repo again in a different local directory and run the commands to drop the database, create the database, and run the migrations?

```
bundle install
rails db:drop
rails db:create
rails db:migrate
rails s
```

Comment (MayPyone) [reply app/views/devise/registrations/new.html.erb:30]: @TedLivist Dear reviewer, I encountered an error when I launched my project again. Thanks for your patience. I will modify my code.

Comment (TedLivist) [reply app/views/devise/registrations/new.html.erb:30]: Alright. That'd be great

**Review by Munsa1 — APPROVED** [review]
Hi @MayPyone, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
App Works As Expected :heavy_check_mark: 
Followed Given Wireframe :heavy_check_mark: 
congrats


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**